### PR TITLE
Add API Surface to get newest available runtimes + SDKs

### PR DESCRIPTION
**Pull request by nagilson**
Resolves https://github.com/dotnet/vscode-dotnet-runtime/issues/764 
* Update; Also adds `recommendedVersion` command to suggest the one newest but active support sdk version for consumers to install. This is not listed on the issue.

 c# Intro

We decided to add an API to allow consumers to tell what newest versions of the SDK and runtime are available.
It's available in both extensions to acquire the list of either. I've also added it to the sample extension, as well as some tests.

We rely on releases-index.json as a data-source: https://github.com/dotnet/core/blob/main/release-notes/releases-index.json

image

image

Also: 
- Adds ability to test only a specific subset of tests (sdk, lib, or runtime) via test lib, test sdk, etc. (See documentation change.)

# Remaining Concerns: 

Note there is a bug in the cache which applies to multiple parts of the extension, including this one; (https://github.com/dotnet/vscode-dotnet-runtime/issues/416) but fixing that should be part of a different PR.

# Questions

I noticed that with async functions in a vscode extension test context and the test framework mocha, the test be marked as failing even if an error is expected in an assert. I initially added a test to ensure an error occurs `The service to request available SDK versions (releases.json) is unavailable.` I removed the test because I couldn't find a way to suppress the error message in the test. https://github.com/dotnet/vscode-dotnet-runtime/pull/791/commits/0f07e17dd88b2413c53496c8276894af21e79a2e#diff-ffab4102fe64ad63fe5d840195be15206110f87171fe39bb79235fa9f56c65c6L117

### CODE REVIEW

File: vscode-dotnet-sdk-extension/package.json [hunk 1]
```diff
@@ -37,9 +37,9 @@
 	"activationEvents": [
 		"onCommand:dotnet-sdk.acquire",
 		"onCommand:dotnet-sdk.acquireStatus",
+		"onCommand:dotnet-sdk.listSdks",
 		"onCommand:dotnet-sdk.uninstallAll",
-		"onCommand:dotnet-sdk.showAcquisitionLog",
-		"onCommand:dotnet-sdk.reportIssue"
```
Comment (nagilson) [inline vscode-dotnet-sdk-extension/package.json:42]: why is this removed?

**Comment by nagilson** [comment]
Something changed in the infrastructure again.. I think the pack and publish failure are totally unrelated to this PR

**Comment by nagilson** [comment]
Going to add the channel-version of this to make our lives easier in another PR.

File: sample/package.json [hunk 2]
```diff
@@ -18,18 +18,6 @@
 	"capabilities": {
 		"virtualWorkspaces": true
 	},
-	"activationEvents": [
```
Comment (nagilson) [inline sample/package.json:21]: In newer versions of VS Code, this is no longer needed, so it's been removed.

File: sample/src/extension.ts [hunk 3]
```diff
@@ -184,6 +188,25 @@ ${stderr}`);
             vscode.window.showErrorMessage((error as Error).toString());
         }
     });
+
+    const sampleSDKListSDKs = vscode.commands.registerCommand('sample.dotnet-sdk.listSdks', async (getRuntimes) => {
```
Comment (nagilson) [inline sample/src/extension.ts:192]: change the name to list available sdks

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 4]
```diff
@@ -139,10 +146,25 @@ export function activate(context: vscode.ExtensionContext, extensionContext?: IE
         }, issueContext(commandContext.errorConfiguration, 'acquireRuntimeStatus'));
         return pathResult;
     });
+
+    const dotnetListSdksRegistration = vscode.commands.registerCommand(`${commandPrefix}.${commandKeys.listNewestSDKs}`,
```
Comment (nagilson) [inline vscode-dotnet-runtime-extension/src/extension.ts]:
```suggestion
    const dotnetListNewestSDKsRegistration = vscode.commands.registerCommand(`${commandPrefix}.${commandKeys.listNewestSDKs}`,
```

File: sample/package.json [hunk 5]
```diff
@@ -73,6 +61,11 @@
 				"title": "Check the status of acquiring the .NET SDK",
 				"category": "Sample"
 			},
+			{
+				"command": "sample.dotnet-sdk.listNewestDotnets",
+				"title": "List the newest available and versions of the .NET SDKs or .NET Runtimes.",
```
Comment (nagilson) [inline sample/package.json:66]:
```suggestion
				"title": "List the newest available versions of the .NET SDK or .NET Runtime.",
```

File: vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+export interface IDotnetListVersionsContext {
+    /**
+     * @remarks
+     * Holds the parameters necessary to acquire the available SDK or Runtimes.
+     * By available, this means all in-support SDKs or Runtimes, and only those at the newest minor version for each major version.
```
Comment (nagilson) [inline vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:10]:
```suggestion
     * By available this means only those at the newest minor version for each major version.
```

**Comment by nagilson** [comment]
The test in mac failing (telemetry) is a known issue. The install not returning a defined result on mac also appears to be not relevant to this PR.

File: vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        const availableVersions : IDotnetListVersionsResult = [];
+        let response = null;
+
+        try
+        {
+            response = await webWorker.getCachedData();
+        }
+        catch(e)
+        {
+            throw new Error(DotnetVersionProvider.dotnetAvailableVersionsPageUnavailableError);
```
Comment (sfoslund) [inline vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts:38]: I don't remember offhand exactly how telemetry works, but these errors seem like the kind of thing you would want to get a count of in telemetry. Have you checked if they're being logged?

**Review by sfoslund — COMMENTED** [review]
This seems like it will be a really useful feature and the changes look great overall! I've left lots of nits and questions, feel free to ping me if you want to discuss anything further :)

File: Documentation/contributing-workflow.md [hunk 8]
```diff
@@ -37,6 +38,11 @@ Note: It is OK to create your PR as "[WIP]" on the upstream repo before the impl
 
 Before making a pull request, be sure to build and test your changes locally with the build script ([windows](https://github.com/dotnet/vscode-dotnet-runtime/blob/main/build.cmd), [mac](https://github.com/dotnet/vscode-dotnet-runtime/blob/main/build.sh)) and test script ([windows](https://github.com/dotnet/vscode-dotnet-runtime/blob/main/test.cmd), [mac](https://github.com/dotnet/vscode-dotnet-runtime/blob/main/test.sh)). To lint your changes, run the test script with the parameter `--tslint`
 
+You can also test only a specific set of tests using the following parameters with the test script:
```
[the remark this answers is not in the document]
Comment (nagilson) [reply Documentation/contributing-workflow.md:41]: Thank you! 😄

File: vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * lts: Long-term support
+ * sts: Standard-term support
+ */
+export type DotnetVersionSupportStatus = 'lts' | 'sts';
+
+/**
+ * @remarks
+ * preview support does not mean the development is stable and actively supported.
+ */
+export type DotnetVersionSupportPhase = 'active' | 'preview' | 'eol' | 'go-live' | 'maintenance';
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:50]: That's a good question. Ideally, we want to discourage people from using EOL dotnets. But: I think we want to expose them because if we don't, it will make it so there's a breaking change in the API every time there's a new EOL SDK. Existing users of the extension seem to hardcode the .NET version they want, so if they rely on this API and then the API suddenly stops providing them a version number, I think that'd be bad. 

I pinged @baronfel about this and he will get back to us on this and the other PM related concerns you mentioned :)

[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:50]: We agreed to expose these for now due to the risk of breaking people

File: vscode-dotnet-sdk-extension/src/test/functional/DotnetCoreAcquisitionExtension.test.ts [hunk 10]
```diff
@@ -238,7 +286,7 @@ suite('DotnetCoreAcquisitionExtension End to End', function() {
 
     // Clean up storage
     await vscode.commands.executeCommand('dotnet-sdk.uninstallAll');
-  }).timeout(600000);
+  }).timeout(maxTimeoutTime * 6);
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-sdk-extension/src/test/functional/DotnetCoreAcquisitionExtension.test.ts:289]: renaming to `standardTimeoutTime`  😆

File: vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+export interface IDotnetListVersionsContext {
+    /**
+     * @remarks
+     * Holds the parameters necessary to acquire the available SDK or Runtimes.
+     * By available this means only those at the newest minor version for each major version.
+     *
+     * @property listRuntimes - If this is not true (false or undefined), list SDK versions available. Elsewise, list runtimes available.
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:12]: That's a good point. The reason is that the Global SDK extension will need to be able to install Runtimes and list the runtimes on the machine. And we're still discussing how to brand the changes (either a new extension, rebrand the runtime one, or rebrand the SDK one.) I talked to Chet and he suggested we keep both in the SDK but not in the runtime extension.

File: vscode-dotnet-runtime-extension/package.json [hunk 12]
```diff
@@ -36,10 +36,10 @@
 	"activationEvents": [
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/package.json:36]: Yes! https://code.visualstudio.com/updates/v1_74#_implicit-activation-events-for-declared-extension-contributions
is the official documentation for that change.

File: vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+import { DotnetVersionSupportStatus, DotnetVersionSupportPhase, IDotnetListVersionsContext, IDotnetListVersionsResult, IDotnetVersion } from '../IDotnetListVersionsContext';
+import { WebRequestWorker } from './WebRequestWorker';
+
+export class DotnetVersionProvider {
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts:8]: I 100% agree. It wasn't really a decision that was made. It's more like I didn't realize that class existed until a few Prs later after this one and never went back to fix it. I will go ahead and do that ;)

File: vscode-dotnet-runtime-extension/src/test/functional/DotnetCoreAcquisitionExtension.test.ts [hunk 14]
```diff
@@ -69,6 +73,30 @@ suite('DotnetCoreAcquisitionExtension End to End', function() {
     assert.include(result!.dotnetPath, context.version);
   }).timeout(40000);
 
+  test('List Sdks & Runtimes', async () => {
+    // A mirror of the test in the SDK Extension.
+    // Test to see if the basic functionality of version acquistion works from both extensions. The more comprehensive test of this code is in the SDK extension.
+
+    const mockWebContext = new MockExtensionContext();
+    const eventStream = new MockEventStream();
+    const webWorker = new MockWebRequestWorker(mockWebContext, eventStream, '', 'MockKey');
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/test/functional/DotnetCoreAcquisitionExtension.test.ts:82]: I love this suggestion, thank you! I will definitely look into it, though I'd imagine it falls fairly low on our priority list for now.

File: vscode-dotnet-runtime-extension/package.json [hunk 15]
```diff
@@ -36,10 +36,10 @@
...
 		"onCommand:dotnet.acquire",
 		"onCommand:dotnet.acquireStatus",
+		"onCommand:dotnet.listNewestDotnets",
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/package.json:39]: Good point, I also don't like the new name with the plural of Dotnets. I asked @baronfel if he could think of a better name.

File: vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:12]: I don't actually think there's a good way to 'hide' the boolean from the runtime extension if one of the extensions (the sdk one) needs both. We could error if you try to acquire SDKs from the runtime extension, but I don't really think that's super helpful, because you'd still see the ability to acquire SDKs from the runtime extension, it just would throw on you. For now I'm going to leave this as-is but feel free to chime in Sarah or @baronfel if you strongly disagree.

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 16]
```diff
@@ -139,10 +146,25 @@ export function activate(context: vscode.ExtensionContext, extensionContext?: IE
         }, issueContext(commandContext.errorConfiguration, 'acquireRuntimeStatus'));
         return pathResult;
     });
+
+    const dotnetListNewestDotnetsRegistration = vscode.commands.registerCommand(`${commandPrefix}.${commandKeys.listNewestDotnets}`,
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/extension.ts:150]: Nope! This was my first time really adding to the code and I did not notice that pattern. Added it, thanks ;)

File: vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+     * @returns
+     * IDotnetListVersionsResult of versions available.
+     *
+     * @throws
+     * Exception if the API service for releases-index.json is unavailable.
+     */
+    async GetAvailableDotnetVersions(commandContext: IDotnetListVersionsContext | undefined, webWorker: WebRequestWorker)
+    {
+        // If shouldObtainSdkVersions === false, get Runtimes. Else, get Sdks.
+        const shouldObtainSdkVersions : boolean = commandContext?.listRuntimes === null || commandContext?.listRuntimes === undefined || !commandContext.listRuntimes;
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts:27]: You are absolutely correct!

File: vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts (hunk 7 above)
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts:38]: They are being logged! Except this error wasn't correctly added to the EventStreams events, nor should it be, since it was duplicated. The telemetry could use a lot of refinement though.

File: vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Utils/DotnetVersionProvider.ts:8]: Done!

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 18]
```diff
@@ -139,10 +146,25 @@ export function activate(context: vscode.ExtensionContext, extensionContext?: IE
...
+        async (commandContext: IDotnetListVersionsContext | undefined, customWebWorker: WebRequestWorker | undefined) => {
+        const webWorker = customWebWorker !== undefined ? customWebWorker : new WebRequestWorker(
+            context.globalState,
+            eventStream,
+            DotnetVersionProvider.availableDontetVersionsUrl,
+            'listSDKVersionsCacheKey'
+        );
+
+        return new DotnetVersionProvider().GetAvailableDotnetVersions(commandContext, webWorker);
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/extension.ts:159]: Good call, though I've simplified the function a lot now so I don't think this is necessary.

File: vscode-dotnet-runtime-extension/package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/package.json:36]: Unfountately the doc they have isnt actually really clear on the distinction you mentioned, but that is what I garner from the underlying subtext and how it seems to work.

File: vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:14]: Ok. We just had a meeting we're going to remove this functionality from the runtime.

File: vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts [hunk 19]
```diff
@@ -23,6 +30,59 @@ export class VersionResolver implements IVersionResolver {
...
+     * @throws
+     * Exception if the API service for releases-index.json is unavailable.
+    */
+    public async GetAvailableDotnetVersions(commandContext: IDotnetListVersionsContext | undefined, webWorker: WebRequestWorker | undefined) : Promise<IDotnetListVersionsResult>
+    {
+        // If shouldObtainSdkVersions === false, get Runtimes. Else, get Sdks.
+        const shouldObtainSdkVersions : boolean = !commandContext?.listRuntimes;
+        const availableVersions : IDotnetListVersionsResult = [];
+
+        const response = webWorker != null ? await webWorker.getCachedData() : await this.webWorker.getCachedData();
```
Comment (nagilson) [inline vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts:53]:
```suggestion
        const response = webWorker ? await webWorker.getCachedData() : await this.webWorker.getCachedData();

```

File: vscode-dotnet-runtime-extension/package.json [hunk 20]
```diff
@@ -36,10 +36,10 @@
 	"activationEvents": [
 		"onCommand:dotnet.acquire",
 		"onCommand:dotnet.acquireStatus",
+		"onCommand:dotnet.listVersions",
```
Comment (sfoslund) [inline vscode-dotnet-runtime-extension/package.json:39]: This line seems to be leftover from previous changes

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 21]
```diff
@@ -50,11 +53,13 @@ namespace configKeys {
 namespace commandKeys {
     export const acquire = 'acquire';
     export const acquireStatus = 'acquireStatus';
+    export const listVersions = 'listVersions'
```
Comment (sfoslund) [inline vscode-dotnet-runtime-extension/src/extension.ts:56]: Leftover from previous changes (also imports on lines 30 and 39)

File: vscode-dotnet-runtime-extension/src/test/functional/DotnetCoreAcquisitionExtension.test.ts [hunk 22]
```diff
@@ -10,10 +10,14 @@ import * as vscode from 'vscode';
 import {
   IDotnetAcquireContext,
   IDotnetAcquireResult,
+  IDotnetListVersionsContext,
+  IDotnetListVersionsResult,
```
Comment (sfoslund) [inline vscode-dotnet-runtime-extension/src/test/functional/DotnetCoreAcquisitionExtension.test.ts:14]: These imports seem to be leftover from previous changes and can be removed

File: vscode-dotnet-runtime-library/src/index.ts [hunk 23]
```diff
@@ -8,16 +8,19 @@ export * from './test/mocks/MockWindowDisplayWorker';
 export * from './test/mocks/MockEnvironmentVariableCollection';
 export * from './IExtensionContext';
 export * from './IDotnetAcquireContext';
+export * from './IDotnetListVersionsContext';
 export * from './IDotnetUninstallContext';
 export * from './IDotnetAcquireResult';
 export * from './IDotnetEnsureDependenciesContext';
 export * from './IExtensionContext';
 export * from './EventStream/EventStreamRegistration';
+export * from './EventStream/IWindowDisplayWorker';
```
Comment (sfoslund) [inline vscode-dotnet-runtime-library/src/index.ts:17]: This is being exported here and imported in the sdk extensions's `extension.ts` but I don't see it ever being used. If that's the case, we should avoid exporting it from this library.

File: vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts [hunk 24]
```diff
@@ -23,6 +30,60 @@ export class VersionResolver implements IVersionResolver {
...
+     * @params
+     * webWorker - This class can use its own web-worker or a custom one for testing purposes.
+     *
+     * @returns
+     * IDotnetListVersionsResult of versions available.
+     *
+     * @throws
+     * Exception if the API service for releases-index.json is unavailable.
+     */
+    public async GetAvailableDotnetVersions(commandContext: IDotnetListVersionsContext | undefined, webWorker: WebRequestWorker | undefined) : Promise<IDotnetListVersionsResult>
```
Comment (sfoslund) [inline vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts:47]: Can we add unit test coverage for this method? It looks like there's an existing VersionResolver unit test class that can be expanded.

Comment (sfoslund) [inline vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts:47]: It seems odd to both create a new `webworker` in the constructor of this class and also accept one as a method parameter here. It looks like you're doing this to follow the pattern of reusing the same `versionResolver` for all the commands being initialized in the sdk extension's `extension.ts` file, which makes sense, but IMO it makes the `VersionResolver` class look a little messy. Is there any reason we need the optional `webWorker` parameter here other than e2e tests?

If we don't need the custom web worker option, we might be able to replace line 53 and some of the parsing logic that follows it with a call to the existing `getReleasesInfo` method. That would also allow us to reuse the existing `ReleasesChannel` type, although it might need to be expanded to save some additional properties that the list command will need. 

If we do need the custom web worker option, IMO it would be cleaner to just create a new `VersionResolver` specifically for this command and pass the custom web request worker in through the constructor, not the method call. 

In terms of the e2e tests, it's not ideal but I don't think it's the end of the world if they just use the real web worker and make real calls to the releases json, especially since that's what the other existing e2e tests already seem to be doing. That's the main differentiator between unit tests and e2e tests anyways- unit tests are made to be run pretty much constantly during development and therefore have to be very granular and super fast, so making network calls wouldn't be acceptable. But I generally see these kinds of e2e tests as "the last line of defense" against bugs, which are run less frequently and therefore there's more room for longer running times, especially if it's exercising more of the production code.

File: vscode-dotnet-runtime-library/src/index.ts (hunk 23 above)
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/index.ts:17]: This is used by the existing setPathEnvVar ;)

File: vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts (hunk 24 above)
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts]: Definitely! I added a pretty basic one, but I think it's an improvement because it tests the online functionality of it. The SDK test is a more comprehensive test.

**Review by sfoslund — APPROVED** [review]
Looks great! One more small comment on the unit tests but otherwise :shipit:

File: vscode-dotnet-runtime-library/src/test/unit/VersionResolver.test.ts [hunk 25]
```diff
@@ -13,6 +14,14 @@ suite('VersionResolver Unit Tests', () => {
     // (prevents us from making web requests in unit tests)
     const resolver: MockVersionResolver = new MockVersionResolver(context, eventStream);
 
+    test('Get Available Versions', async () => {
+        const result : IDotnetListVersionsResult = await resolver.GetAvailableDotnetVersions(undefined);
+        // Assert that the call gives a result with members. The data may change, so we did not include data here.
+        // Comprehensive e2e tests with data are in the sdk extension.
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/test/unit/VersionResolver.test.ts:20]: You're right, thank you :) I went ahead and added that

**Comment by nagilson** [comment]
It's extremely kind of you to review, thank you Sarah. :) 
@JL03-Yue is out on bootcamp for a while, so I may bypass the branch policy this time so we can get it in on time.

File: sample/src/extension.ts [hunk 26]
```diff
@@ -34,6 +38,7 @@ export function activate(context: vscode.ExtensionContext) {
     runtimeExtension.activate(context);
     sdkExtension.activate(context);
 
+
```
Comment (Forgind) [inline sample/src/extension.ts:41]: nit: extra space

File: sample/src/extension.ts [hunk 27]
```diff
@@ -184,6 +195,33 @@ ${stderr}`);
             vscode.window.showErrorMessage((error as Error).toString());
         }
     });
+
+    const sampleSDKlistVersions = vscode.commands.registerCommand('sample.dotnet-sdk.listVersions', async (getRuntimes : boolean) => {
+        if (!getRuntimes) {
+            getRuntimes = JSON.parse(await vscode.window.showInputBox({
+                placeHolder: 'false',
+                value: 'false',
+                prompt: 'Acquire Runtimes? Use `true` if so, else, give `false`.',
```
Comment (Forgind) [inline sample/src/extension.ts:204]: This looks like a sample, so maybe the answer should be 'no', but should this be localized?

File: vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts [hunk 28]
```diff
@@ -0,0 +1,54 @@
+/* --------------------------------------------------------------------------------------------
```
Comment (Forgind) [inline vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:1]: Is this the right copyright header? Looks wrong to me.

**Review by Forgind — APPROVED** [review]
I definitely know what I'm doing.

Actually, nagilson just told me to approve 😉

File: sample/src/extension.ts (hunk 27 above)
Comment (nagilson) [reply sample/src/extension.ts:204]: It'd be good to have localization for everything in this repo, but currrently most of these strings aren't

File: vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts (hunk 28 above)
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/IDotnetListVersionsContext.ts:1]: That's a very good point. There's an issue for it here: https://github.com/dotnet/vscode-dotnet-runtime/issues/905. every file has it, so Ill fix it in a different pr

**Review by marcpopMSFT — APPROVED** [review]

**Review by yimmee123 — COMMENTED** [review]
Thk
